### PR TITLE
add custom f1 score metric to the researcher tutorial

### PR DESCRIPTION
**Pull request by haifeng-jin**

### CODE REVIEW

File: guides/intro_to_keras_for_researchers.py [hunk 1]
```diff
@@ -567,6 +567,59 @@ def call(self, inputs):
 by calling `layer.reset_metrics()` on any layer or model.
 """
 
+import tensorflow as tf
+from tensorflow import keras
```
Comment (fchollet) [inline guides/intro_to_keras_for_researchers.py:571]: They're imported earlier already

File: guides/intro_to_keras_for_researchers.py [hunk 2]
```diff
@@ -567,6 +567,59 @@ def call(self, inputs):
...
+    self.true_positives = self.add_weight(name='tp', initializer='zeros')
+    self.false_positives = self.add_weight(name='fp', initializer='zeros')
+    self.false_negatives = self.add_weight(name='fn', initializer='zeros')
+
+  def update_state(self, y_true, y_pred, sample_weight=None):
+    y_pred = tf.math.greater_equal(y_pred, self.threshold)
+    y_true = tf.cast(y_true, tf.bool)
+    y_pred = tf.cast(y_pred, tf.bool)
+
+    true_positives = tf.logical_and(tf.equal(y_true, True), tf.equal(y_pred, True))
```
Comment (fchollet) [inline guides/intro_to_keras_for_researchers.py:587]: They're bool already, so no need for the `tf.equal`, no?

File: guides/intro_to_keras_for_researchers.py [hunk 3]
```diff
@@ -567,6 +567,59 @@ def call(self, inputs):
...
+    false_positives = tf.logical_and(tf.equal(y_true, False), tf.equal(y_pred, True))
+    false_negatives = tf.logical_and(tf.equal(y_true, True), tf.equal(y_pred, False))
+
+    true_positives = tf.cast(true_positives, self.dtype)
```
Comment (fchollet) [inline guides/intro_to_keras_for_researchers.py:591]: `self.dtype` has never been set or defined, prefer using `'float32'`

File: guides/intro_to_keras_for_researchers.py [hunk 4]
```diff
@@ -567,6 +567,59 @@ def call(self, inputs):
...
 
+import tensorflow as tf
+from tensorflow import keras
+
+class F1Score(keras.metrics.Metric):
+
+  def __init__(self, name='f1_score', threshold=0.5, **kwargs):
+    super().__init__(name=name, **kwargs)
+    self.threshold = 0.5
+    self.true_positives = self.add_weight(name='tp', initializer='zeros')
```
Comment (fchollet) [inline guides/intro_to_keras_for_researchers.py:578]: Need a dtype for the weights

File: guides/intro_to_keras_for_researchers.py [hunk 5]
```diff
@@ -567,6 +567,59 @@ def call(self, inputs):
...
+    false_positives = tf.logical_and(tf.equal(y_true, False), tf.equal(y_pred, True))
+    false_negatives = tf.logical_and(tf.equal(y_true, True), tf.equal(y_pred, False))
+
+    true_positives = tf.cast(true_positives, self.dtype)
+    false_positives = tf.cast(false_positives, self.dtype)
+    false_negatives = tf.cast(false_negatives, self.dtype)
+
+    if sample_weight is not None:
+      sample_weight = tf.cast(sample_weight, self.dtype)
+      true_positives = tf.multiply(true_positives, sample_weight)
```
Comment (fchollet) [inline guides/intro_to_keras_for_researchers.py:597]: Prefer `*`, the code should look natural, numpy-like

File: guides/intro_to_keras_for_researchers.py [hunk 6]
```diff
@@ -567,6 +567,59 @@ def call(self, inputs):
...
+      false_positives = tf.multiply(false_positives, sample_weight)
+      false_negatives = tf.multiply(false_negatives, sample_weight)
+
+    self.true_positives.assign_add(tf.reduce_sum(true_positives))
+    self.false_positives.assign_add(tf.reduce_sum(false_positives))
+    self.false_negatives.assign_add(tf.reduce_sum(false_negatives))
+
+  def result(self):
+    precision = tf.math.divide_no_nan(self.true_positives, self.true_positives + self.false_positives)
```
Comment (fchollet) [inline guides/intro_to_keras_for_researchers.py:606]: Prefer `/`

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the draft PR!

**Review by fchollet — APPROVED** [review]
LGTM, thank you! (I pushed some copyedits)
